contracts/modules/maker/MakerV2Loan.sol

```diff
@@ -0,0 +1,642 @@
+
+contract JugLike {
+    function drip(bytes32) external;
+}
+
+interface IUniswapFactory {
+    function getExchange(address _token) external view returns(IUniswapExchange);
+}
+
+interface IUniswapExchange {
+    function getEthToTokenOutputPrice(uint256 _tokensBought) external view returns (uint256);
+    function getEthToTokenInputPrice(uint256 _ethSold) external view returns (uint256);
+    function getTokenToEthOutputPrice(uint256 _ethBought) external view returns (uint256);
+    function getTokenToEthInputPrice(uint256 _tokensSold) external view returns (uint256);
+}
+
+/**
+ * @title MakerV2Loan
+ * @dev Module to migrate old CDPs and open and manage new vaults. The vaults managed by
+ * this module are directly owned by the module. This is to prevent a compromised wallet owner
+ * from being able to use `TransferManager.callContract()` to transfer ownership of a vault
+ * (a type of asset NOT protected by a wallet's daily limit) to another account.
+ * @author Olivier VDB - <olivier@argent.xyz>
+ */
+contract MakerV2Loan is Loan, MakerV2Base {
+
+    // The address of the MKR token
+    GemLike internal mkrToken;
+    // The address of the WETH token
+    GemLike internal wethToken;
+    // The address of the WETH Adapter
+    JoinLike internal wethJoin;
+    // The address of the Jug
+    JugLike internal jug;
+    // The address of the Vault Manager (referred to as 'CdpManager' to match Maker's naming)
+    ManagerLike internal cdpManager;
+    // The address of the SCD Tub
+    SaiTubLike internal tub;
+    // The Maker Registry in which all supported collateral tokens and their adapters are stored
+    MakerRegistry internal makerRegistry;
+    // The Uniswap Exchange contract for DAI
+    IUniswapExchange internal daiUniswap;
+    // The Uniswap Exchange contract for MKR
+    IUniswapExchange internal mkrUniswap;
+    // Mapping [wallet][ilk] -> loanId, that keeps track of cdp owners
+    // while also enforcing a maximum of one loan per token (ilk) and per wallet
+    // (which will make future upgrades of the module easier)
+    mapping(address => mapping(bytes32 => bytes32)) public loanIds;
+
+    // Mock token address for ETH
+    address constant internal ETH_TOKEN_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    // Method signatures to reduce gas cost at depoyment
+    bytes4 constant internal ERC20_TRANSFER = bytes4(keccak256("transfer(address,uint256)"));
+    bytes4 constant internal WETH_DEPOSIT = bytes4(keccak256("deposit()"));
+    bytes4 constant internal WETH_WITHDRAW = bytes4(keccak256("withdraw(uint256)"));
+    bytes4 constant internal VAT_HOPE = bytes4(keccak256("hope(address)"));
+    bytes4 constant internal CDP_GIVE = bytes4(keccak256("give(uint256,address)"));
+    bytes4 constant internal TUB_GIVE = bytes4(keccak256("give(bytes32,address)"));
+    bytes4 constant internal ETH_TOKEN_SWAP_OUTPUT = bytes4(keccak256("ethToTokenSwapOutput(uint256,uint256)"));
```

Comment:
These are used in `abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...)` which is faster to do with `abi.encodeWithSignature(string memory signature, ...)`

https://solidity.readthedocs.io/en/latest/units-and-global-variables.html#abi-encoding-and-decoding-functions

Reply:
We made that choice in the past because we had measured a reduction of deployment gas cost for `--runs=200` when storing a single `constant bytes4 ` to be used by multiple `abi.encodeWithSelector` instead of inlining multiple times the same string signature in `abi.encodeWithSignature`. Interestingly, I redid the comparison and I now find that though our approach was indeed cheaper for `--runs=200`, using `abi.encodeWithSignature` is actually cheaper for `--runs=999`, even when inlining the signature multiple times in the code. I'll make the change to use `abi.encodeWithSignature` everywhere.